demo.py

```diff
@@ -0,0 +1,214 @@
+                "hisat2_index": f"{ref_prefix}/reference/hisat2.tar.gz",
+                "salmon_index": f"{ref_prefix}/reference/salmon.tar.gz",
+                "rsem_index": f"{ref_prefix}/reference/rsem.tar.gz",
+                "skip_bbsplit": False,
+                "pseudo_aligner": "salmon",
+                "umitools_bc_pattern": "NNNN",
+            },
+        )
+
+    def synindex_info(self) -> LaunchInfo:
+        """Generate LaunchInfo for nf-synindex."""
+        rnaseq_info = self.rnaseq_info()
+        assert rnaseq_info.params
+
+        return LaunchInfo(
+            run_name=self.get_run_name("synindex"),
+            pipeline="Sage-Bionetworks-Workflows/nf-synindex",
+            revision="main",
+            profiles=["sage"],
+            params={
+                "s3_prefix": rnaseq_info.params["outdir"],
+                "parent_id": self.output_folder,
+            },
+            workspace_secrets=["SYNAPSE_AUTH_TOKEN"],
+        )
+
+
+class TowerRnaseqFlow(FlowSpec):
+    """Flow for processing RNA-seq dataset on Tower."""
+
+    tower = NextflowTowerOps()
+    synapse = SynapseOps()
+    s3 = s3fs.S3FileSystem()
+
+    # Synapse ID for a YAML file that describes an RNA-seq dataset
+    # Here's an example:
+    #   id: my_test_dataset
+    #   samplesheet: syn51514475
+    #   output_folder: syn51514559
+    dataset_id = Parameter("dataset_id", help="Dataset Synapse ID", type=str)
+
+    def monitor_workflow(self, workflow_id):
+        """Monitor any workflow run (wait until done)."""
+        monitor_coro = self.tower.monitor_workflow(workflow_id)
+        status = asyncio.run(monitor_coro)
+        if not status.is_successful:
+            message = f"Workflow did not complete successfully ({status})."
+            raise RuntimeError(message)
+        return status
+
+    @step
+    def start(self):
+        """Entry point."""
+        self.next(self.load_dataset)
+
+    @step
+    def load_dataset(self):
+        """Load dataset from Synapse YAML file."""
+        with self.synapse.fs.open(self.dataset_id, "r") as fp:
+            kwargs = yaml.safe_load(fp)
+        self.dataset = RnaseqDataset(**kwargs)
+        self.next(self.transfer_samplesheet_to_s3)
+
+    @step
+    def transfer_samplesheet_to_s3(self):
+        """Transfer raw samplesheet from Synapse to Tower S3 bucket."""
+        sheet_uri = f"{TMPDIR_PREFIX}/{self.dataset.id}.csv"
+        sheet_contents = self.synapse.fs.readtext(self.dataset.samplesheet)
+        self.s3.write_text(sheet_uri, sheet_contents)
+        self.dataset.samplesheet = sheet_uri
+        self.next(self.launch_synstage)
+
+    @step
+    def launch_synstage(self):
+        """Launch nf-synstage to stage Synapse files in samplesheet."""
+        launch_info = self.dataset.synstage_info()
+        self.synstage_id = self.tower.launch_workflow(launch_info, "spot")
+        self.next(self.monitor_synstage)
+
+    @step
+    def monitor_synstage(self):
```

Comment:
Just out of curiosity, is there a "sensor" concept in metaflow?

Reply:
Not natively, but we could simulate this behavior using retries ([docs](https://docs.metaflow.org/api/step-decorators/retry)). I got this idea from the [Prefect docs](https://discourse.prefect.io/t/is-there-an-equivalent-to-sensors-in-prefect-how-do-i-trigger-event-driven-workflows/76#h-1-state-based-execution-with-signals-1). For example, we can fail the step whenever the workflow isn't complete yet and retry every 30 minutes. If we want to allow the workflow up to one week to complete, we would set the number of retries to 336 (= 7 x 24 / 0.5).